baseplate/clients/redis.py

```
@@ -108,12 +170,134 @@ def report_runtime_metrics(self, batch: metrics.Client) -> None:
         batch.gauge("pool.in_use").replace(in_use)
         batch.gauge("pool.open_and_available").replace(open_connections - in_use)
 
-    def make_object_for_context(self, name: str, span: Span) -> "MonitoredRedisConnection":
-        return MonitoredRedisConnection(name, span, self.connection_pool)
+        retry_stats = RetryStats.get_retry_stats(self.connection_pool)
+        for event_name, value in retry_stats.items():
+            batch.counter(event_name).increment(value)
+
+    def make_object_for_context(
+        self, name: str, span: Span
+    ) -> "MonitoredRedisConnection":
+        kwargs = {}
+        if hasattr(self.connection_pool, "redis_enable_retries"):
+            kwargs["enable_retries"] = getattr(
+                self.connection_pool, "redis_enable_retries"
+            )
+        return MonitoredRedisConnection(name, span, self.connection_pool, **kwargs)
+
+
+class PipelineWithRetry(Pipeline):
+    """Subclass of Pipeline that will automatically retry pipelines with only read commands."""
+
+    def __init__(  # type: ignore
+        self, connection_pool, response_callbacks, transaction, shard_hint, retriable
+    ) -> None:
+        super().__init__(
+            connection_pool=connection_pool,
+            response_callbacks=response_callbacks,
+            transaction=transaction,
+            shard_hint=shard_hint,
+        )
+        self.is_retriable_pipeline = retriable
+
+    def _execute_pipeline(self, connection, commands, raise_on_error):  # type: ignore
+        retries = 0
+        MAX_RETRIES = 10
+        MAX_RETRY_WAIT_MS = 50
+
+        cmd_list = [args[0] for args, _ in commands]
+        is_read_only = all(cmd in RETRIABLE_COMMANDS for cmd in cmd_list)
+
+        while retries < MAX_RETRIES:
+            try:
+                return super()._execute_pipeline(connection, commands, raise_on_error)
+            except Exception:
```

I think I'd be hesitant to catch all exceptions here. You probably only want to retry certain failures, right? For example, if I have a bug in my code and I try to `SUNION` on a non-set filed, I'll get an error like the following:
```
127.0.0.1:6379> SET foo 5
OK
127.0.0.1:6379> SUNION foo
(error) WRONGTYPE Operation against a key holding the wrong kind of value
```

That's not going to work no matter how many times I retry.

---

that command won't get retried, because it's not considered a retriable command. I'm happy to switch that to just catching the same exceptions we filter for the non-pipeline case, but it feels silly because the client will only raise the first exception within the pipeline, so filtering doesn't do much for you.

---

The problem with catching `Exception` is that if there's a legit bug (e.g. out-of-bounds indexing an array) we catch that too. For library code, that's almost never the right answer. So we should definitely find a more tightly scoped catch here. It doesn't have to check subtypes like in the other path but not catching the base Exception is important.